Feat/factory

Grouflon opened the pull request:
@berdal84 Je viens de rajouter un système de factory pour instantier des objets directement depuis le TypeDesc. Je pense qu'y a une grosse problématique au niveau de comment pouvoir sélectionner la méthode d'allocation parceque j'imagine que tout le monde veut pas allouer ses objets avec des new, mais je suis assez inexpérimenté en allocateurs et je sais pas vraiment quelle serait la bonne manière de faire ça.

Sinon je pense que c'est super cool, le serializer peut maintenant instantier et sérialiser des objets de manière générique juste en ajoutant une metadata OwnedPointer sur ton pointeur réfléchi.

Grouflon committed:
First implementation of generic factories

Grouflon committed:
updated readme

Grouflon committed:
Update README.md

berdal84 commented on README.md:
```diff
@@ -3,8 +3,9 @@ mirror is a lightweight C++ reflection framework that aims at providing a simple
 
 ## Manifesto (enforced in this framework, but actually applies to any coding project)
 - Simple ideas should be simple to write.
+- If you are not using it, it should not do anything.
```
C'est bien ça.

berdal84 commented on README.md:
```diff
@@ -77,7 +80,7 @@ MIRROR_ENUM(MyEnum)
 
 ## How to use
-- Any reflected class gains a public `GetClass()` static function that allow to iterate through reflected members, access their types and find their address on given instances. You can also access it from the oustide with the function `mirror::GetClass<T>()`
+- Any reflected class gains a public `GetClass()` static function that allow to iterate through reflected members, access their types and find their address on given instances. You can also access the reflected type one any type from the oustide with the function `mirror::GetTypeDesc<T>()` or `mirror::GetTypeDesc(myVariable)`
```
one any type => on any type ?

berdal84 commented on mirror_base.cpp:
```diff
@@ -160,18 +162,54 @@ namespace mirror
+		{
+			return m_factory();
+		}
 		return nullptr;
```
je mettrai presque une assert la. genre: did you check if `hasFactory()` ?

Grouflon replied:
Je suis plus trop sûr de ce genre d'asserts. Le travail me fait douter, je me ramollis...
Non mais ça peut peut-être être valide pour le user de checker son pointeur à la sortie et de faire un truc, j'avoue je sais pas trop encore

berdal84 commented on mirror_base.h:
```diff
@@ -15,6 +15,7 @@ namespace mirror
 {
 	class Class;
 	class TypeDesc;
+	typedef void* (*FactoryFunctionPtr)();
```
as tu envisagé std::function ? pourquoi ne pas l'utiliser ?

Grouflon replied:
quel est l'avantage par rapport au pointeur en l'occurence ?

berdal84 replied:
je pense que je pensais au fait de typer par autre chose que void*, mais en fait toi tu veux une fonction generique donc aucun avantage en l'occurence.

berdal84 commented on mirror_base.h:
```diff
@@ -47,17 +61,25 @@ namespace mirror
 		size_t getTypeHash() const { return m_typeHash; }
+		size_t getSize() const { return m_size; }
+
+		// @TODO(2021/02/15|Remi): Allow the user to choose their allocator
```
j'en sais pas plus que toi sur les allocateurs, mais regardes du coté des implem de std::shared_ptr ou unique_ptr pour voir comment ils font.

Grouflon replied:
pas con

berdal84 commented on tools/BinarySerializer.cpp:
```diff
@@ -265,12 +265,65 @@ namespace mirror
+					}
+					else if (m_isReading)
+					{
+						// @TODO(2021/02/15|Remi): May leak the previous value of the pointer. What should we do ? Whose responsibility is it ?
```
en mode lecture, si le pointeur n'est pas nullptr pourquoi ne pas conserver l'instance et écraser ses membres plutot que creer une instance ? ca resouds le pb de responsabilite (niveau memoire du moins).

Grouflon replied:
la j'ai pas encore implémenté les pointeurs "virtuels", mais quand ce sera le cas, je serai pas sur que le pointeur sera effectivement de la même classe.
Cela dit je pourrai le tester aussi, mais peut-être que ça va rendre le pipeline d'init d'un objet un peu complexe non ? genre certains objets auront droit à leur constructeur, et d'autres seront serialisés sur un état déjà existant...
vasy je sais pas, c'est difficile de concevoir des trucs abstraits comme ça je trouve

Grouflon replied:
Mais c'est une bonne idée d'un point de vue perfs

berdal84 commented on mirror_macros.h:
```diff
@@ -56,6 +58,12 @@ public:\
+// @NOTE(2021/02/15|Remi): Wrapping this code into curly brackets throw out a compilation error, I don't get why.
+// Not a big deal since so far we only need one but still...
+#define MIRROR_FACTORY()\
+	static auto s_factory = []() -> void* { return new classType(); };\
```
pourquoi void* en retour ? c'est pour la genericite des fct de serialisation ?

Grouflon replied:
ouais, y aucun autre type que je peux retourner des méthodes `instantiate()` des `TypeDesc` je pense

berdal84 replied:
y'a un truc qui pourrait t'interesser: `std::any`, jette un oeil, ca permet d'interroger le type avec un truc genre: `std::hold_alternative<T>(std::any) bool` (de memoire)

berdal84 reviewed (commented):
voila j'ai fait une petite revue. j'espere que ca aidera. tchu

Grouflon commented:
Merci beaucoup, c'est cool !

Grouflon committed:
added a virtual type wrapper to allow calls to final types

Grouflon committed:
Types must now be initialized in the source file with the corresponding macro
Can export to DLL
Introduced TypeID type instead of size_t
removed dependency to std::string
